### PR TITLE
Generalize bootloader

### PR DESCRIPTION
**Pull request by hathach**
This PR generalize this repo making it cross-platform uf2 with sufficient enough board API (for now)
https://github.com/adafruit/uf2-esp32s/blob/generalize-bootloader/src/board_api.h

and tested on following platforms with successful firmware.uf2 update to circuipython 
- esp32s2 : saola 1r
- mimxrt10xx: imxrt1060_evk
- stm32f4: feather_stm32f405_express, stm32f411ve_discovery

additional features include
- Indicator with LED and RGB (neopixel)
- Double reset tap to DFU
- Optional pritnf debug log with UART or SWD

Although the major feature is UF2 update and double tap works, there is still lots of work for improvement and portings. However, I feel it is good enough for review and merge, so that we could test, improve and submit pr to the project e.g hid, webusb, cf2, self-update, screen, docs etc ...

@ladyada @tannewt we will need to have a new name for this project and rename it accordingly, let's me know which name you prefer.

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+    tud_task();
+  }
+#endif
+}
+
+//--------------------------------------------------------------------+
+// Device callbacks
+//--------------------------------------------------------------------+
+
+// Invoked when device is mounted
```
Comment (tannewt) [inline src/main.c:146]: Could we say "plugged in" and "unplugged"? "mounted" is confusing because it doesn't relate to the mass storage drive.

**Review by tannewt — COMMENTED** [review]
Overall it looks really good! I still like the TinyUF2 name. We could rename the existing repo and then rename this one to it I think.

File: src/main.c (hunk 1 above)
Comment (hathach) [reply src/main.c:146]: done

**Comment by hathach** [comment]
> Overall it looks really good! I still like the TinyUF2 name. We could rename the existing repo and then rename this one to it I think.

I have delete the other adafruit/tinyuf2 repo fork, since we didn't make any development on that repo so far. once this is merged, we could rename it

**Comment by hathach** [comment]
@tannewt let me know if you want to make any other changes. It is better to merge this before working on other feaures..

**Review by tannewt — APPROVED** [review]
No other changes from me! Are you losing any existing functionality? (In particular, the APA102 support that @UnexpectedMaker added.) If not, then feel free to merge.

**Comment by hathach** [comment]
> No other changes from me! Are you losing any existing functionality? (In particular, the APA102 support that @UnexpectedMaker added.) If not, then feel free to merge.

I did a manual merge, it should still work as it is. Though @UnexpectedMaker may want to test it again since I don't have that board.

**Comment by hathach** [comment]
merged and also rename the repo to **tinyuf2**

**Comment by ladyada** [comment]
yayyy!

**Comment by UnexpectedMaker** [comment]
Great job @hathach ! Going to try this out later today :)

**Comment by UnexpectedMaker** [comment]
Seems to be working fine here!

**Pull request merged**